src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -5119,23 +5145,47 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
             }
             else
             {
-                TypeWithState rightType;
+                TypeWithState rightState;
                 if (!node.IsRef)
                 {
-                    rightType = VisitOptionalImplicitConversion(right, leftLValueType, UseLegacyWarnings(left, leftLValueType), trackMembers: true, AssignmentKind.Assignment);
+                    rightState = VisitOptionalImplicitConversion(right, leftLValueType, UseLegacyWarnings(left, leftLValueType), trackMembers: true, AssignmentKind.Assignment);
                 }
                 else
                 {
-                    rightType = VisitRefExpression(right, leftLValueType);
+                    rightState = VisitRefExpression(right, leftLValueType);
                 }
 
-                TrackNullableStateForAssignment(right, leftLValueType, MakeSlot(left), rightType, MakeSlot(right));
-                SetResult(node, TypeWithState.Create(leftLValueType.Type, rightType.State), leftLValueType);
+                // If the LHS has annotations, we perform an additional check for nullable value types
+                CheckDisallowedNullAssignment(rightState, leftAnnotations, right.Syntax.Location);
+
+                TrackNullableStateForAssignment(right, leftLValueType, MakeSlot(left), rightState, MakeSlot(right));
+                SetResult(node, TypeWithState.Create(leftLValueType.Type, rightState.State), leftLValueType);
             }
 
             return null;
         }
 
+        private FlowAnalysisAnnotations GetLValueAnnotations(BoundExpression expr)
+        {
+            var symbol = expr switch
+            {
+                BoundPropertyAccess property => property.PropertySymbol,
+                BoundIndexerAccess indexer => indexer.Indexer,
+                _ => null
+            };
+
+            var annotations = symbol switch
+            {
+                PropertySymbol property => lastParameter(property.SetMethod)?.FlowAnalysisAnnotations ?? FlowAnalysisAnnotations.None,
+                _ => FlowAnalysisAnnotations.None
+            };
+
+            return annotations & (FlowAnalysisAnnotations.DisallowNull | FlowAnalysisAnnotations.AllowNull);
+
+            static ParameterSymbol lastParameter(MethodSymbol method)
+                => method is null || method.Parameters.IsDefault ? null : method.Parameters.Last();
```

Comment:
>method.Parameters.IsDefault [](start = 37, length = 27)

Is this ever true?

Reply:
It does occur (in `CS0655ERR_BadNamedAttributeArgumentType_Dynamic`). I'll take a deeper look. Feels like `Parameters` should always return `Empty` rather than `default`.

Reply:
Turns out getting `default` from `Parameters` was a symptom of a cycle. 
Fixed in the latest commit (using the same pattern you used to avoid such cycles with attributes).